url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/fastme/tree.java

```diff
@@ -0,0 +1,1616 @@
+		   			else
+		   				updateSubTreeAverages(A, par, v, direction.RIGHT);
+		   			A[left.head.index][par.head.index]
+		   			     = A[par.head.index][left.head.index]
+		   			     = (left.bottomsize*A[left.head.index][par.head.index]
+		   			     + A[v.index][par.head.index]) / (left.bottomsize + 1);
+		   		}
+		   		break;    
+		   	case RIGHT: //point of insertion is below the edge right
+		   		//1-distant average
+		   		A[e.head.index][e.head.index] = 
+		   			(e.bottomsize*A[e.head.index][e.head.index] + 
+		   			A[v.index][e.head.index])/(e.bottomsize + 1);        
+		   		//2-distant averages
+		   		A[e.head.index][e.tail.index] = 
+		   			A[e.tail.index][e.head.index] = 
+		   			(e.bottomsize*A[e.head.index][e.tail.index] + 
+		   			A[v.index][e.tail.index])/(e.bottomsize + 1);  
+		   		A[left.head.index][right.head.index] = 
+		   			A[right.head.index][left.head.index] = 
+		   			(right.bottomsize*A[right.head.index][left.head.index]
+		   			+ A[left.head.index][v.index]) / (right.bottomsize+1);
+		   		//3-distant avereages involving right
+		   		if (null != sib)
+		   		{
+		   			updateSubTreeAverages(A, sib, v, direction.SKEW);
+		   			A[right.head.index][sib.head.index]
+		   			     = A[sib.head.index][right.head.index]
+		   			     = (right.bottomsize*A[right.head.index][sib.head.index]
+		   			     + A[sib.head.index][v.index]) / (right.bottomsize + 1);
+		   		}
+		   		if (null != par)
+		   		{
+		   			if (e.tail.leftEdge == e)
+		   				updateSubTreeAverages(A, par, v, direction.LEFT);
+		   			else
+		   				updateSubTreeAverages(A, par, v, direction.RIGHT);
+		   			A[right.head.index][par.head.index]
+		   			     = A[par.head.index][right.head.index]
+		   			     = (right.bottomsize*A[right.head.index][par.head.index]
+		   			     + A[v.index][par.head.index]) / (right.bottomsize + 1);
+		   		}
+		   		break;
+		}
+	}
+	private void GMEsplitEdge( node v, edge e, double[][] A )
+	{
+	  String nodelabel;
+	  String edgelabel;
+	  edge newPendantEdge;
+	  edge newInternalEdge;
+	  node newNode;
+	    
+	  nodelabel = "I"+(this.size+1); 
+	  newNode = new node( nodelabel, null, this.size + 1 );  
+	  edgelabel = "E"+this.size; 
+	  newPendantEdge = new edge( edgelabel, newNode, v, 0.0 );   
+	  edgelabel = "E"+(this.size+1); 
+	  newInternalEdge = new edge( edgelabel, newNode, e.head, 0.0 );     
+	  //update the matrix of average distances
+	  //also updates the bottomsize, topsize fields
+	  
+	  GMEupdateAveragesMatrix(A,e,v,newNode);
+
+	  newNode.parentEdge = e;
+	  e.head.parentEdge = newInternalEdge;
+	  v.parentEdge = newPendantEdge;
+	  e.head = newNode;
+	  
+	  this.size = this.size + 2;
+
+	  if (e.tail.leftEdge == e) 
+	    {
+	      newNode.leftEdge = newInternalEdge;
+	      newNode.rightEdge = newPendantEdge;
+	    }
+	  else
+	    {
+	      newNode.leftEdge = newInternalEdge;
+	      newNode.rightEdge = newPendantEdge;
+	    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 494:503, 680:690.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.